src/main/resources/jupyter/edit-mode.js

```diff
@@ -0,0 +1,337 @@
+    }
+
+    async function getLock() {
+        const payload = {
+            ...basePayload,
+            method: 'POST',
+            body: JSON.stringify(getLocalPath())
+        }
+
+        fetch(lockUrl, payload)
+            .then(res => {
+                handleLockStatus(res)
+                return res.json()
+            })
+            .catch(err => {
+                console.error(err)
+            })
+    }
+
+    async function toggleMetaFailureBanner(shouldShow) {
+        const bannerId = "notification_metaFailure"
+        const bannerText = "Failed to check notebook status, changes may not be saved to workspace. Retrying..."
+
+        removeElementById(bannerId)
+
+        if (shouldShow) {
+            const bannerStyling = "btn btn-xs navbar-btn btn-danger"
+
+            $('#notification_area').append(
+                $('<div>').attr({
+                    "id": bannerId,
+                    "class": bannerStyling
+                })
+                .append($('<span>').html('<i class="fa fa-exclamation-triangle"></i> ' + bannerText))
+            );
+        }
+
+    }
+
+    async function handleLockStatus(res) {
+        if (!res.ok) {
+            const status = res.status
+            const errorText = res.statusText
+
+            if (status == 409) {
+                res.json().then(body => {
+                    const message = getLockConflictBody(body.lockedBy)
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => {
+        return `<p>This file is currently being editted by ${lockHolder}.</p>` +
+            `<br/><p>You can make a copy, or run it in Playground Mode to explore and execute its contents without saving any changes.`;
+    }
+
+    async function promptUserWithModal(title, buttons, htmlBody) {
+        if (modalOpen) return
+
+        modalOpen = true
+
+        dialog.modal({
+                body: $('<p>').html(htmlBody),
+                title: title,
+                buttons: buttons,
+                notebook: Jupyter.notebook,
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
+
+        console.log('switching to playground path: ', newPath)
+
+        const payload = {
+            ...basePayload,
+            method: 'PATCH',
```

Comment:
What are we patching here? Can you add a comment? The semantics I'd expect here are:

- copy or "save as" to playground mode directory
- redirect to new file

Reply:
added

Reply:
Per offline discussion: want copy rather than move, here and below (save as)